### PR TITLE
feat(multi-tenancy) migrate multi-tenancy code to daikon

### PR DESCRIPTION
**Pull request by bguillon-td**
**What is the problem this Pull Request is trying to solve?**

Multi-tenancy code should be shared across applications and services.
 
**What is the chosen solution to this problem?**

Take multi-tenancy from IAM / other apps and move it here and document it
 
**Link to the JIRA issue**
 
**Please check if the Pull Request fulfills these requirements**
- [x] The PR commit message follows our [guidelines](../CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request
 
**[ ] This Pull Request introduces a breaking change**

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/21/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/21/)

File: daikon-multitenant/multitenant-spring-async/src/main/java/org/talend/daikon/multitenant/async/ContextPropagator.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// %InstallDIR%\features\org.talend.rcp.branding.%PRODUCTNAME%\%PRODUCTNAME%license.txt
+//
+// You should have received a copy of the agreement
+// along with this program; if not, write to Talend SA
+// 9 rue Pages 92150 Suresnes, France
+//
+// ============================================================================
+package org.talend.daikon.multitenant.async;
+
+public interface ContextPropagator {
```
Comment (sgandon) [inline daikon-multitenant/multitenant-spring-async/src/main/java/org/talend/daikon/multitenant/async/ContextPropagator.java]: Could you add some javadoc on the public interfaces please.

File: daikon-multitenant/multitenant-spring-async/src/test/java/org/talend/daikon/multitenant/async/MultiTenantApplication.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @Component
+    public static class MessageQueue {
+
+        private final LinkedBlockingQueue<Message> messages = new LinkedBlockingQueue<>(1);
+
+        @Async
+        public void publish(String content) {
+            System.out.println("Hello");
```
Comment (sgandon) [inline daikon-multitenant/multitenant-spring-async/src/test/java/org/talend/daikon/multitenant/async/MultiTenantApplication.java]: Could you remove this please, may be use a logger.

File: daikon-multitenant/multitenant-core/src/main/java/org/talend/daikon/multitenant/provider/DefaultTenant.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * 
+ * @author Clint Morgan (Tasktop Technologies Inc.)
+ */
+public class DefaultTenant implements Tenant {
+
+    private static final long serialVersionUID = 1L;
+
+    private Object identity;
+
+    private Object data;
```
Comment (sgandon) [inline daikon-multitenant/multitenant-core/src/main/java/org/talend/daikon/multitenant/provider/DefaultTenant.java]: What is this data used for ? I dont see any use of it.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/22/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/22/)

**Comment by sgandon** [comment]
a quick question, I can see in the PR 2 tenant identification strategies, url and header.
I suppose we have our how implementation with regard to IAM with a resolution of the tenant id using the bearer token. I that so or am I missing something ?

**Comment by bguillon-td** [comment]
Yes that's right, we have our own implementation in a private repository

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/23/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/23/)

**Pull request merged**